### ISSUE TITLE
RFC: Adoption of ``std::mdspan`` like structures in C++ code

### ISSUE THREAD

**Issue by Kai-Striega** [scipy/scipy#20334]
Hello all,

Disclaimer: I'm not a C++ expert, this is an issue to track the discussion amongst people who are, instead of leaving it distributed amongst several issues.

With the increase in C++ in SciPy, we have several methods for handling arrays being introduced. For example, https://github.com/scipy/scipy/pull/20320 introduces an implementation of ``std::mdspan`` and https://github.com/scipy/scipy/pull/19970 introduces its own (simple) version of an array.

I'd like to propose a standard way of handling arrays in C++ code, or, at a minimum, a recommended way. I quite like the ``std::mdspan`` idea. This was standardised in C++ 23 (which we haven't adopted yet) https://github.com/scipy/scipy/pull/20320 introduces a backward compatible version of this, as does CuPy's [CArray](https://github.com/cupy/cupy/blob/028889eef3ba110829d677726348aa0f75aadb4e/cupy/_core/include/cupy/carray.cuh#L379).

What are everyone's thoughts on this?

**Comment by ev-br** [comment]
First and foremost, IMO this RFC is not only for C++ experts! It may impact us mere mortals as well.

With the C++ momentum by @izaid and @steppi in scipy.special and other people, I think it'd be great to
- formulate a best-practices guide for C++ in SciPy, and
- put common building blocks to _lib or _build_utils for all submodules to use.

At least three things come to mind:
- how to link LAPACK from C++ (this is gh-20002)
- how to properly pass complex-valued arrays from python to C++ and back (scipy.special developed some means? but it's awkward to use in other subpackages)
- how to grab a view onto a numpy array from C++.

For the latter, `std::mdspan` seems to be the recommended way. So my suggestion would be to put the backcompat implementation from gh-20320 to _build_utils or some such. One other thing from past experience: we need  a way to toggle bounds checking at compile time, this saves long hours of debugging.
So an `access_policy` with bounds checking would be a great addition to what is being introduced in gh-20230.

One other relatively minor sticky point in C++ vs python glue is signed / unsigned indices: C++ containers use size_t while python indexing is `ssize_t` (I believe?). A best-practices recommendation of how to convert from one to the other would be great.

**Comment by izaid** [comment]
Hi all! As has been mentioned, @steppi and I are on a long-standing quest to greatly improve SciPy's special. Our goal is to get all of the numerical functions in special currently mixed between C, C++, Cython, Python, and F̶o̶r̶t̶r̶a̶n̶ into a single header-only C++ library. That library will then be able to provide special function implementations to all the other array libraries, and indeed it already is being used in CuPy. We are doing this by converting the special function implementations into NumPy ufuncs with C++ kernels. There are some tools we need to do this (and those will exit SciPy when this header-only library becomes its own submodule), so another question is what SciPy wants in general.

As part of our efforts, we've come across all the things discussed above. I'm happy to help if useful. Some thoughts are below.

For complex numbers, we use `std::complex<float>` and `std::complex<double>`. You can directly cast a pointer to a `npy_cfloat` and a `npy_cdouble` to them, then operate on them in a convenient fashion.

The other thing that came up is how to handle functions that operate on subarrays. In our case, this arose from functions like the associated Legendre polynomials that return a `(m + 1, n + 1)` array for each value `z` in the input. Really all that is needed is a simple class that wraps a pointer, a shape, and strides, no memory ownership and no reference counting. One can either write their own, or use the upcoming standard `std::mdspan` from C++23 that is exactly intended for this use case. CuPy is an example of a library that wrote their own, the `CArray` above, but they did that before `std::mdspan` was accepted into the standard.

Note that to do views without memory allocation, the number of dimensions must be known at compile-time. I don't think this is a great restriction, as most functions in SciPy that need to drop to C++ somehow know if they are operating on 1D, 2D, 3D, or whatever arrays.

Most of the special work has been tools to help with ufuncs, I think what's needed here is actually way simpler. Again, happy to help if that is useful.

**Comment by izaid** [comment]
> That's the sales pitch but not that the arrays must be C-order contiguous or strided etc hence we need to babysit everything just as we do for regular C flatbuffers. Otherwise you have to copy things to fresh arrays. So this problem is not solved by C++, this is already the default behavior. All it brings is the `md` part over the regular `span`. So it will be as annoying as Cython memoryviews or malloc'ed arrays to convert back and forth between NumPy arrays.
> 
> Hence my not so enthusiastic comments about this new class. It is accepted to standard but still miles away from a NumPy or Fortran like convenience and indexing/slicing support and safely manipulating them. To be honest, I don't even know what the benefit is **for us** compared to regular std::`span`.

Sure, that's definitely why it's good to have a discussion! I'm here mostly to provide information about what we've done in special.

I will note that `std:mdspan` is meant to support different layouts, and it already has compile-time customization for row-major (rightmost contiguous), column-major (leftmost contiguous), and arbitrary strides. If you want something more exotic, it can also be customized. It's also been [backported](https://github.com/NVIDIA/cccl) to CUDA by NVIDIA.

It's primary purpose is to establish a standard pattern for views on multidimensional arrays in C++. One can always create their own thing instead.

**Comment by ilayn** [comment]
Yes sure, I'm not holding you accountable or ambassador of C++ 😃 It is just what it is.

**Comment by izaid** [comment]
> Yes sure, I'm not holding you accountable or ambassador of C++ 😃 It is just what it is.

Personally, I do think it's the right way to go. These things are helping a lot in special, so perhaps they can also help elsewhere.

**Comment by tylerjereddy** [comment]
In case it is relevant, I know the National Labs here implemented https://github.com/kokkos/mdspan, and the [PyKokkos](https://github.com/kokkos/pykokkos) Python bindings to Kokkos are still being developed at UT Austin (I was helping them implement ufuncs before I ran out of bandwith).

**Comment by izaid** [comment]
> In case it is relevant, I know the National Labs here implemented https://github.com/kokkos/mdspan, and the [PyKokkos](https://github.com/kokkos/pykokkos) Python bindings to Kokkos are still being developed at UT Austin (I was helping them implement ufuncs before I ran out of bandwith).

Not only is that relevant, the Kokkos repository is actually what we're using in special!

**Comment by Kai-Striega** [comment]
Kokko's/CUDA's implementations seem worth looking at, does anyone know whether their licenses are compatible with SciPy's?

**Comment by fancidev** [comment]
> Our goal is to get all of the numerical functions in special currently mixed between C, C++, Cython, Python, and F̶o̶r̶t̶r̶a̶n̶ into a single header-only C++ library.

May I ask how is this going to be related to  those special math functions already available in `boost`?

**Comment by izaid** [comment]
> > Our goal is to get all of the numerical functions in special currently mixed between C, C++, Cython, Python, and F̶o̶r̶t̶r̶a̶n̶ into a single header-only C++ library.
> 
> May I ask how is this going to be related to those special math functions already available in `boost`?

We'll use Boost where we can, but Boost doesn't support CUDA among other things. SciPy has an enormous collection of special functions, they are valuable in and of themselves. We discussed this several months ago, see https://github.com/scipy/scipy/issues/19404

**Comment by steppi** [comment]
Also, @fancidev see #19601 and https://github.com/cupy/cupy/pull/8140 for examples of this in action. I’d say we can’t use Boost directly because it doesn’t support CUDA. The hope is to have a single special function library that could be used by SciPy, CuPy, PyTorch, Jax, and other array libraries.

**Comment by fancidev** [comment]
Thanks both for the reply and references! I find [this comment](https://github.com/scipy/scipy/issues/19404#issuecomment-1775431903) quite informative and answers all my questions. Looking forward to have those functions reusable in C++ (even if by copy pasting)!

**Comment by ilayn** [comment]
> Personally, I do think it's the right way to go. These things are helping a lot in special, so perhaps they can also help elsewhere.

I have nothing against C++ as it is being used in many parts, I would not use for anything but that's anectodal. 

But if I consider numerical work, I am not really convinced about C++/F90 in terms of portability. So far there are two, that stood the test of time; one is C and the other is F77. Both C++ and F90 are standing on the laurels of their "less well-mannered cousins". The rest is just lots of some successful and some broken codebases with mixed feelings. I know Sandia labs went all in with C++ which is perfectly fine for them with their team fully behind that decision. A lot of others went with Modern Fortran, whatever that means today. But we are not a C++ codebase. We just need to get by with maintainable code + performance if possible. Much to my astonishment after so many decades, Reference LAPACK is trying to get some C++ code into the majestic Fortran codebase that we all love with their home-made matrix classes https://github.com/Reference-LAPACK/lapack/pull/991 So nobody is agreeing on a proper array structure and we are in 2024. So this makes no sense to me and has all the signs of sunken cost for the future. 

For linear algebra work, currently mdspan does not really offer anything yet if compared to, say even in C++ ecosystem, Eigen, and that's not smooth sailing at all [this page](https://eigen.tuxfamily.org/dox/group__TopicUnalignedArrayAssert.html) being my favorite. 

For `special` it might be OK since there is not that many array operation requirements but if you want to do a bit more tricky ops then `mdspan` is just an unnecessary abstraction that will continuously get in the way. 

However there is [std::linalg](https://en.cppreference.com/w/cpp/numeric/linalg) is in the works (C++26 and onwards) and who knows when we will be able to switch to it, if we ever stay with C++. Then things can get really nice if it works.

But if I accept that level of complexity, I would definitely switch to Rust instead which is already much better than this. So not sure where the benefit of `mdspan` lies. If you don't want the whole bulky Eigen dependency and still use mdspan locally inside `special` that's obviously a different requirement for keeping a lean header only module. I totally get that.

Like we discussed in the very early times on another RfC, C++ stuff is only OK in my opinion if we stay away from the bad parts of C++ and these data structures look like the bad parts of C++ to me with unnecessary abstractions without the actual convenience we need that Fortran or NumPy (or even Eigen) offers.

If what @steppi said about making special a separate codebase and hydrating all stakeholders from a common place, is going to happen soon, then noone probably will care what you did in the bowels of the library and just compile it. But switching to mdspan overall in SciPy seems to me a really bad investment regardless what the standard committee claims.

**Comment by tylerjereddy** [comment]
> I would definitely switch to Rust instead

+1, but I'm afraid compiler/ecosystem support remains a blocker for the foreseeable future? I had a great experience improving perf with Rust at https://github.com/scipy/scipy/issues/14719. I'm with you on the pain of C++, to me it is like the saying about regular expressions where you then have two problems, the actual problem and dealing with C++.

That said, if folks are willing to maintain it I guess I've mostly stayed out of the way. The C++ adoption seems to be happening in NumPy too. We used to have a rule about order of preference being this for maintenance reasons: Python, Cython, C, C++, Fortran. I believe that was in the SciPy `1.0` paper, but then there were various discussions around allowing C++ more frequently.

**Comment by ilayn** [comment]
> +1, but I'm afraid compiler/ecosystem support remains a blocker for the foreseeable future?

Very good point, though we are also in the tough spot for switching to anything newer than C++17 but do you have particular architectures you have in mind? I'm not as versed as you are regarding different architectures but I tried already with a few embedded targets with preliminary success. 

> I had a great experience improving perf with Rust at https://github.com/scipy/scipy/issues/14719.

Oh how did I miss that? That's a wonderful result and also gives me more confidence that I am not the only one getting affirmative results on the "Fearless Concurrency" claims

**Comment by h-vetinari** [comment]
> > +1, but I'm afraid compiler/ecosystem support remains a blocker for the foreseeable future?
> 
> Very good point, though we are also in the tough spot for switching to anything newer than C++17 but do you have particular architectures in mind?

I regularly come back to wanting to have a NEP29-style document for our platform/arch/compiler support, perhaps like the tiered approach in Cpython or rust.

In any case, if Rust is good enough for the kernel, I have a hard time imagining that it couldn't be used in SciPy; C support is certainly broader, but are there actually any architectures these days with a C++ compiler that's unavailable for rust?

Aside from the unclear outcome with the various standardisation efforts, we're also signing up for a long wait before we can require C++20 (much less C++23/26), so while I'm sympathetic to the overall effort being proposed here, I do see the argument by @ilayn too, and tend to prefer we'd go for rust.

**Comment by lucascolley** [comment]
I think the important question now is then: would converting from C++ (using similar features to the current work) to Rust be significantly more work than converting directly from C / Cython?

**Comment by h-vetinari** [comment]
Every rewrite takes a substantial effort, so converting first to C++ and later to Rust certainly needs more effort than one rewrite.

That said, the one argument I see for the C++ side is GPU support, not least since Nvidia is strongly betting on C++. I don't know what Rust's story for heterogeneous computing is, but given that it's all LLVM under the hood, I kinda expect it to be... not that bad (certainly given the time we'd wait for universal support of newer C++ standards)?

**Comment by fancidev** [comment]
After reading the great discussions, it seems the main debate is whether the complexity of C++ justifies its adoption in SciPy.

It appears to me the key question is whether the C++ code to be introduced will be used exclusively by SciPy, or do you foresee it to be adopted by other projects? The latter goal is certainly much more ambitious, more complex, and more demanding, and for that goal C++ may be a suitable choice because of its wide adoption in industries.

**Comment by steppi** [comment]
> If what @steppi said about making special a separate codebase and hydrating all stakeholders from a common place, is going to happen soon, then noone probably will care what you did in the bowels of the library and just compile it. But switching to mdspan overall in SciPy seems to me a really bad investment regardless what the standard committee claims.

I think it depends to an extent on whether our EOSS 6 proposal is funded, otherwise it might be more difficult to secure the bandwidth, but I'd hope to put it in a separate codebase within one year, and and start doing outreach to get more stakeholders interested within 6 months.

At the moment I don't feel qualified to have an opinion on the questions of Rust vs future versions of C++ vs <???> and how exactly we should standardize on ways to work with arrays in low level code, but plan to take some time to educate myself on these matters.

**Comment by h-vetinari** [comment]
> or do you foresee it to be adopted by other projects

I'll let @steppi answer himself, though widespread reusability is certainly a goal AFAICT (both in terms of libraries that can leverage this, as well as architectures that this can run on, in particular GPUs).

I think no matter the implementation choice, we realistically need to provide the API in multiple flavours (either through a FFI or wrappers), e.g. C-style, C++-style, and _potentially_ Rust-style (depending on what the ecosystem ends up doing, but certainly if the implementation itself were to be in Rust).

**Comment by ev-br** [comment]
> compiler/ecosystem support remains a blocker for the foreseeable future?

So if we drop the kokkos implementation somewhere into scipy, just as gh-20230 does, do we expect issues from the compile/build side? 

Realistically, I don't think we could/should/would maintain it in any meaningful sense, it'd be just there as a black box. Would that cause problem?

**Comment by izaid** [comment]
It's been really interesting to watch this conversation. And I think the point about the Linux kernel adopting Rust is a good one.

I also think it's worthwhile to consider the **different** topics that have brought up in this thread. The original post was "_can we have a standard way of handling arrays in C++ code_". Then there was some discussion about "_Rust is pretty good, can we use it instead_". Rather than get into the points about C++ versus Rust, perhaps there should be some serious thought about how to get the **best practices of each language** for the whole project. It's also worth noting that Python extension modules are written in C, hence there is a natural connection to the C / C++ ecosystem.

In a very real way, it seems SciPy has always been a multilanguage project: C, C++, Fortran, Python, maybe Rust too at some point. And that also evolves with time. For various good reasons, Fortran is now being taken out of the codebase. Maybe at some point Rust will become a large part of it. Maybe SciPy will even branch off from just Python, like how IPython became Jupyter. In the meantime, I would think the point isn't choosing C++ versus Rust, but writing the best C++ or the best Rust or the best whatever.

So, back to the original question. I would think the need for a data structure that reinterprets memory as a multidimensional array is clear. That's exactly what NumPy is for Python. This data structure really only needs to do the memory reinterpretation, which basically means providing indexing for different memory layouts or making subviews. If you are working in C++, you will need something in C++. If you are working in Rust, you will need something in Rust. If we don't provide this, people are forced to mix up things like pointer indexing with the actual numerical algorithms. That has been one of the problems in special.

So what is the solution for C++? Either write your own or use what has now been standardised. The benefits of using something standardised now is it'll be right there in the upcoming standard and it can interact with other code in the ecosystem. A potentially good way to go would be to work out best practices for C++, but that wouldn't mean C++ necessarily just appears all over the ecosystem. I guess it would still have to be discussed case by case. A separate issue could discuss working out best practices for Rust and potential adoption.

**Comment by izaid** [comment]
> > compiler/ecosystem support remains a blocker for the foreseeable future?
> 
> So if we drop the kokkos implementation somewhere into scipy, just as [gh-20230](https://github.com/scipy/scipy/issues/20230) does, do we expect issues from the compile/build side?
> 
> Realistically, I don't think we could/should/would maintain it in any meaningful sense, it'd be just there as a black box. Would that cause problem?

It is backported to C++14. It has the same issues as other third-party code, you are reliant on someone else having done a good job! To me, it seems quite good. Also to NVIDIA, who have adopted it in their CUDA standard library.

The other option is to just write your own class like the CuPy `CArray` mentioned earlier, in which case you are just creating the same thing (which is not necessarily bad, but that's what it is).

**Comment by ilayn** [comment]
>  I would think the point isn't choosing C++ versus Rust, but writing the best C++ or the best Rust or the best whatever.

This is my main argument indeed that C++ after 40 years has no best practices. Just lots of people saying the others are doing it wrong and defining newer and newer things. Note that Eigen is also a standard tool de facto accepted by everyone. So is all in between implementations for span and vector and so on. The language is void of a proper array structure that's why I think we should not accept experimental features of C++ even though they are backported, compilers are the important detail not the standard.

**Comment by fancidev** [comment]
Out of curiosity, is there a procedure in SciPy developers to resolve design choices if they fail to arrive at a consensus? Like a vote or something?

**Comment by tylerjereddy** [comment]
> Out of curiosity, is there a procedure in SciPy developers to resolve design choices if they fail to arrive at a consensus? Like a vote or something?

Technically we have BDFL, though he's inactive at the moment. Seems to be voting-based mostly, though usually seems to reach an informal agreement between devs before it gets as far as actually needing to count votes. This may help a bit: https://scipy.github.io/devdocs/dev/governance.html

**Comment by andyfaff** [comment]
As an outsider to some aspects of the discussion - in many of my own projects I let cython deal with all the hard work of multidimensional array wrangling (strides, etc). After all, it's a superset of Python so the syntax is pretty easy. Then I drop down into C/C++ for the actual scalar kernel computation once all the hard work has been done.

From past experience the more is in Python/cython (and less in C) the easier it is to maintain. e.g. With the TNC optimiser code the extension interface was originally written in C, calling into vendored C code. IMO writing Python extensions directly in C is painful, lots of weird function names (e.g. `PyArg_ParseTuple`) to understand, etc. This made it hard to maintain. Rewriting the interface in cython made it much more maintainable and easier to extend in the future.

**Comment by rgommers** [comment]
A few thoughts on the actual proposal here:
- Vendoring a well-tested thing like `std::mdspan` (leaving aside the licensing issue I commented on in gh-20320) seems clearly preferable to me to reinventing the wheel here,
- It's a lot of code, and we don't have a great history with directly vendoring code and then keeping it up to date; we may want to stick this in a separate git submodule for traceability. If the current `mdspan.h` is the result of running `make_single_header.py` from https://github.com/kokkos/mdspan (is it?), that could even be done dynamically to avoid committing anything that's not directly from upstream.

**Comment by rgommers** [comment]
Re the preferred programming languages discussion: in most cases the answer is "it depends". There are pros and cons to everything. The most clear change we've had over the last 1-2 years: new Fortran 77 code is no longer acceptable. Beyond that, there are trade-offs. C++ seems like the best of a not-so-great lot at least for `special`. 

Cython (or Pythran) are the right tool in certain circumstances where pure Python doesn't cut it, but certainly not always. They're easier to maintain for not-too-complex code for the average SciPy maintainer. And Cython itself is fairly well-maintained. However, we also know there are significant downsides to Cython: too large binaries, long compile times, a lot of extra complexity in the build system (and poor build system support in general), fused types and const support are not very mature, niche language so little general-purpose tooling support, and the code generation for our Cython bindings and exported `cython_special/blas/lapack` APIs are some of  the most complex and difficult to understand/modify parts of the code base we have. I'm pretty sure that the notes in http://scipy.github.io/devdocs/dev/roadmap-detailed.html#use-of-cython are still correct; Cython isn't really scalable to the size of the whole SciPy code base.

tl;dr it depends

**Comment by ev-br** [comment]
Re: cython vs C++ : I think we could do with a worked example. scipy.special gives one (and an excellent one at that!). However, it's somewhat special : it seems one can use numpy (g)ufuncs + scalar kernels. What would be helpful I think is to take a cython extension, port it to C or C++, including python bindings. It would be interesting to see what the binary size would be if we go from programming in Cython to programming in C/C++ and only wrapping in Cython. Does this solve the binary size issue? I don't know and I doubt it can be just guessed. 

Likewise for Rust: if we are to seriously consider it, let's first look at a concrete example.

**Comment by izaid** [comment]
> Re: cython vs C++ : I think we could do with a worked example. scipy.special gives one (and an excellent one at that!). However, it's somewhat special : it seems one can use numpy (g)ufuncs + scalar kernels. What would be helpful I think is to take a cython extension, port it to C or C++, including python bindings. It would be interesting to see what the binary size would be if we go from programming in Cython to programming in C/C++ and only wrapping in Cython. Does this solve the binary size issue? I don't know and I doubt it can be just guessed.
> 
> Likewise for Rust: if we are to seriously consider it, let's first look at a concrete example.

I think this is a very good point! In the case of special, things are very well-defined. All the C++ functions are (g)ufunc kernels, and literally all the language bridging is done by (g)ufunc calls. That is exactly the whole point of the special work, to get things into exactly that model.

For other situations where things don't neatly fit into the (g)ufunc model, it's a very good question. I personally try to stay in pure Python as much as I can,  but acknowledge there are cases where that doesn't work.

**Comment by rgommers** [comment]
> It would be interesting to see what the binary size would be if we go from programming in Cython to programming in C/C++ and only wrapping in Cython. Does this solve the binary size issue? I don't know and I doubt it can be just guessed.

We have enough examples of code that was converted, or will be converted I think. I'm pretty sure that the answer is "no". E.g., see https://github.com/scipy/scipy/issues/16583 for bindings to individual functions in `boost.math`. It's better size-wise than having everything in Cython (e.g., see the `cdflib` size explosion after Cythonization in https://github.com/scipy/scipy/pull/19560#issuecomment-1854727303), but it's still quite bad. Depends on what one is doing so hard to give a general rule, but the penalty of using Cython seems to be anywhere from 1.2x to 50x.

The `cython_special` extension module is probably the best-case scenario, lots of functions in a single module that ends up being quite large, so the constant overhead per extension module that Cython adds is small. It then ends up being ~1.6 MB for `_ufuncs.so` + `_ufuncs_cxx.so` (the functionality) and ~1.9 MB for `cython_special.so`. So the bindings take up 300 kb or so, 15-20% of the total size. (all sizes are on-disk sizes in release build mode)

The `stats._boost` case is an example of what happens with smaller extension modules (the more common case): it's 1.45 MB for 9 ufuncs - which is larger than the 266 ufuncs contained in `special._ufuncs`. It's not exactly apples to apples, but it's close enough - and it ends up being about a 45x size difference per ufunc.

**Comment by ilayn** [comment]
OK fine, fine, let me rewrite `cdflib` in C. Jeez, such a tough crowd 😄 

Note that much of the bloat, I feel like, coming from using Cython as a substitue for `pybind11` or `nanobind` or whatever, that is, using it as C bridge. I think we should be more willing to go the extra mile and get these integrated through a proper C API instead of generating binaries through transpiling with Cython overhead added on top of them. Cython itself does not have too much to add but templating and using it as we do, a quick-dirty safeguard around C code probably does.

For exposing C code probably we can do better but don't ask me how. I'm already regretting every step, heh.

**Comment by ev-br** [comment]
> let me rewrite cdflib in C

There's a stub: https://github.com/ev-br/scipy/tree/cpp_cdflib  :-)

**Comment by ev-br** [comment]
> the penalty of using Cython seems to be anywhere from 1.2x to 50x.

Compated to what, the C extension without wrappers, manual C API wrap, pybind11?

**Comment by ilayn** [comment]
> > let me rewrite cdflib in C
> 
> There's a stub: https://github.com/ev-br/scipy/tree/cpp_cdflib :-)

I want to use this so badly. But unfortunately we fixed so many bugs after this, all of them are worse than looking for needle in a needlestack so I can't even risk copy pasting anymore 😞

**Comment by rgommers** [comment]
> Compated to what, the C extension without wrappers, manual C API wrap, pybind11?

Any of those - it really varies strongly case by case. Here is another existing example: in `scipy/ndimage` we have a `_ctest.c` (using the Python and NumPy C APIs directly) and a `_cytest.pyx`, which are both about the same length (179 vs 153 LoC) and implement the same functions. The end result is `_ctest.so` is 16 kb, `_cytest.so` is 86 kb - so a 5.4x difference. Compile time is 270 ms for `.c`, 4.1 s for `.pyx` - a 15x difference.

Back to the original topic:
1. The `special` PR that triggered this issue/discussion is merged. I think we're good with using `mdspan` in `special`. 
2. For other modules/code, I think it's still a case-by-case decision. Sometimes C++ is the right tool, sometimes Cython, sometimes C. Not sure how much more we can say here now. We can probably close this issue.

**Comment by lucascolley** [comment]
I think there are still a few relevant points about writing down best practices in https://github.com/scipy/scipy/issues/20334#issuecomment-2020528186 that could be useful, but maybe more appropriate to open a new issue if those are still wanted.

**Comment by rgommers** [comment]
> I think there are still a few relevant points about writing down best practices in [#20334 (comment)](https://github.com/scipy/scipy/issues/20334#issuecomment-2020528186) that could be useful, but maybe more appropriate to open a new issue if those are still wanted.

I'd suggest copying those bullet points in a next comment here, and for each one just link to code that does the thing in what we think is a "best practice" way. I don't see much point in a new issue asking to write a whole "how to use C++ guide" - that's both a ton of work and a large bike shed.

**Comment by sturlamolden** [comment]
I am one of those who is partially responsible for C++ code in SciPy. I would like to apologize for that. 

For the sake of portability, ease of coding, I should have used C. 

I am sorry. 

C++ has with the latest revolutions of the language turned into a nightmare.

As for multidimensional arrays in native code, NumPy C API has solved that a long time ago. I would like to encourage the use of it instead of some modern C++.

**Comment by sturlamolden** [comment]
> But if I consider numerical work, I am not really convinced about C++/F90 in terms of portability. So far there are two, that stood the test of time; one is C and the other is F77. Both C++ and F90 are standing on the laurels of their "less well-mannered cousins". 

Yes, that seems to be the case, though C++98 and F90 were not bad, it was the later revolutions of C++ and Fortran that screwed things up. 

C++ got bad post C++11.

Fortran got bad with the OOP in F2003. Who does OOP in Fortran?

**Comment by steppi** [comment]
Thanks for weighing in @sturlamolden. Please keep in mind that the discussion here is not whether modern C++ is a good or bad language, but "for situations where the maintainers working on a part of the codebase decide C++ is the best tool for the job, should we standardize on using `mdspan` to manage `n`-dimensional arrays?" The goal here is to avoid proliferation of competing array implementations in different parts of SciPy due to lack of coordination between the maintainers of different submodules. Compared to the alternative of using `mdspan` in `special` (#20320) , a different custom implementation in `interpolate` (#19970), and perhaps something completely different elsewhere, it seems like a good idea to work out in advance what we consider best practices for working with ndarrays in C++.

**Comment by sturlamolden** [comment]
@steppi I think you misunderstand. The NumPy C API is adequate even when using C++. std::mdspan is a poor replacement. We use the NumPy C API a lot anyway (or at least we used to).

**Comment by sturlamolden** [comment]
I have looked at your examples, e.g. #20320. It seems you have encountered some bad examples. NumPy C API should not be used directly in Cython. In C++ we can wrap the functionality we need in a class. If you need to see how to do it, look in Python scripting for comoutational science by Hans Petter Langtangen.

**Comment by steppi** [comment]
> I have looked at your examples, e.g. #20320. It seems you have encountered some bad examples. NumPy C API should not be used directly in Cython. In C++ we can wrap the functionality we need in a class. If you need to see how to do it, look in Python scripting for computational science by Hans Petter Langtangen.

@izaid and I both know how to use the NumPy C API. Using it in `special` isn't an option for the work we are doing, because we seek to make things Array API backend agnostic, so that these functions can be used in CuPy, PyTorch, etc.

**Comment by sturlamolden** [comment]
Looking at #20320, you can convert ```z``` to a NumPy array in Cython and then pass that to C++. You do not need to pass anything else than PyArrayObject* to C++ in order to support CuPy and PyTorch. Those array type conversions are just boilerplate code you do in Python or Cython.

**Comment by izaid** [comment]
I think @steppi is explaining things well, but I just want to be very clear about one thing. The purpose of using `std::mdspan` is to be able to have a (non-owning) multidimensional array view that can be used on the CPU, on the GPU, on device-independent code. The NumPy C API cannot be used except on the CPU. So our options are, as explained before, either write our own custom structure or use something standard, which is `std::mdspan`.

In `scipy.special`, there are kernels that need to handle subarrays, so we need the multidimensional indexing capability. This is enabling us to bring many special functions to CuPy that it didn't have before. And that's just the start.

**Comment by ilayn** [comment]
I think these points will only get more diverging over time if we don't do the separation in this version or next so that you can also focus things in a more refined way in your own pace. I do acknowledge that CuPy and NVIDIA and others went C++ way and doing their own thing that's truly great. I can only be happy about it. I do have to admit that I am extremely cynical when it comes to big corporations tipping their toes into open source but I don't have a dog in this race at all. 

But on the other hand, please also take note that C++ has significant problems with its feature inclusion and the standard definition. `mdspan` is supposed to be a nD-`memoryview` Cython has been providing since a decade already. Hence it is not just old folks shouting at the sun. This has been happening in the last 15 years many many times inside its `std::bloat` and there are good reasons people hate it with passion. While I am not that emotional about it, you don't need to be a hacker to see the argument on that side too.

However, coming back to this, the RfC is about using it inside SciPy code and it sounds like we don't have enough use cases to decide either way. I guess since the `scikit-special` or whatever the name would be splitting we don't need to have an argument after all.

**Comment by steppi** [comment]
> I think these points will only get more diverging over time if we don't do the separation in this version or next so that you can also focus things in a more refined way in your own pace.

The split is in the works! @izaid and would like to try to get this C++ code into a separate repo within the SciPy org in time for 1.14.

For what we're doing in `special`, C++ seems to be, if not the only, at least the only reasonable option for writing device independent code. NVIDIA has carved out a path we can take using C++, the work required is already immense, and I don't think it could even be accomplished if done in a less efficient way (from a development perspective). For having to rely on a proprietary compute platform, I think it just is what it is; this is the work that needs to be done to have array API compatibility for `special`.

> However, coming back to this, the RfC is about using it inside SciPy code and it sounds like we don't have enough use cases to decide either way. I guess since the `scikit-special` or whatever the name would be splitting we don't need to have an argument after all.

Yeah, this RFC is about using `mdspan` in SciPy proper, and the examples in `special` only show that that there can be use cases in niche circumstances. So far, I think @ev-br is the only one who has started to tinker with using `mdspan` in SciPy proper, and I don't think he's come to any kind of conclusion yet. We'll have to see where that goes.

**Comment by sturlamolden** [comment]
In SciPy proper I am responsible for the C++ code in cKDTree. That is, I could have used C and nobody would have objected… 

Anyhow, I originally wrote it using NumPy C API in C++. Now it seems to use a combination of NumPy C API and Cython memoryviews in Cython, but all of the PyArrayObject* stuff is taken out of C++. In doing so it does no longer support arbitrary strides, but indexing looks nicer without the strides perhaps. I do not any longer remember who did it and why. But the use of memoryvews in Cython has certainly made the C++ code less flexible. And combining NumPy C API and memoryviews in Cython is messy. Also we do not need memoryviews just to extract pointers from an ndarray. 

Personally I do not think std::mdspan would improve on anything in cKDTree. I would just be another layer of complication in something C++ and Cython has made too messy already. The code unfortunately uses three ways of representing an ndarray already, and std::mdspan will mean a fourth way, not a replacement for the three we already have. It is one more step on the path to unmaintainable code. Future maintainers might not understand why things are the way they are. Originally I only used ONE way to represent an ndarray. Now we have three, and this RFC is suggesting to add another. I am -1 on that. 

If I were to write cKDTree all over again (and perhaps I will…), I would not have used C++ but plain C. I would not have been equally fast (say, 90 % of the current), but ut would not have been bloated and much more easy to maintain. The only thing we get from C++ that is not in C is std::nth_element, but implementing introselect in C is not rocket science.

**Comment by steppi** [comment]
Sorry @sturlamolden, I should have mentioned that the suggestion is to use `mdspan` in new code, in lieu of implementing a comparable custom array view, precisely in such situations where a maintainer has a use case for such a thing. I think we’re in agreement that updating existing C++ like cKDTree to use `mdspan` would just be churn with no real benefit.

**Comment by sturlamolden** [comment]
> Sorry @sturlamolden, I should have mentioned that the suggestion is to use `mdspan` in new code, in lieu of implementing a comparable custom array view, precisely in such situations where a maintainer has a use case for such a thing. I think we’re in agreement that updating existing C++ like cKDTree to use `mdspan` would just be churn with no real benefit.

It would have a benefit, it it were done consistently and the other three means of representing an ndarray were cleared out. The C++ code in cKDTree originally had only one way of representing an ndarry.

**Comment by rgommers** [comment]
The simpler mdspan-like version of @izaid is in, and the vendored copy can be added. This issue can be closed as resolved I think - we can go ahead with use of `std::mdspan`.

**Comment by Kai-Striega** [comment]
Thanks all for your inputs.
